tools/atf-kill-tree.1
.\"
.\" Automated Testing Framework (atf)
.\"
.\" Copyright (c) 2008 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this
.\"    software must display the following acknowledgement:
.\"        This product includes software developed by the NetBSD
.\"        Foundation, Inc. and its contributors.
.\" 4. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND
.\" CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\" GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
.\" IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
.\" IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd January 14, 2008
.Dt ATF-KILL-TREE 1
.Os
.Sh NAME
.Nm atf-kill-tree
.Nd kills a process tree
.Sh SYNOPSIS
.Nm
.Ar pid
.Nm
.Fl h
.Sh DESCRIPTION
.Nm
is a utility that recursively kills all children of a given process;
in other words, it kills a process tree.
The algorithm used iterates over the tree from top to bottom.
For each process, it sends a
.Sq SIGSTOP
to it to freeze its execution, then looks for all of its live children,
stores this list and at last sends the requested signal to the process
followed by a
.Sq SIGCONT
to resume its execution.
(In some platforms this last step is done the other way around, though.)
This process is later restarted on each of the children processes.
.Pp
In the first synopsis form,
.Nm
proceeds to kill the given process and all of its children.
.Pp
In the second synopsis form,
.Nm
will print information about all supported options and their purpose.
.Pp
The following options are available:
.Bl -tag -width flag
.It Fl h
Shows a short summary of all available options and their purpose.
.It Fl s Ar signal
Specifies the (numerical) signal to be sent.
Defaults to
.Sq SIGTERM .
.El
.Sh SEE ALSO
.Xr atf 7


NEWS
                     Automated Testing Framework (atf)
                        The NetBSD Foundation, Inc.
---------------------------------------------------------------------------


Version 0.4
===========

Release date: NOT RELEASED YET; STILL UNDER DEVELOPMENT
Status:       Experimental

Changes:

* Added two new manual pages, atf-c++-api and atf-sh-api, describing the
  C++ and POSIX shell interfaces used to write test programs.

* Added a pkg-config file, useful to get the flags to build against the
  C++ library or to easily detect the presence of ATF.

* Added a way for test cases to require a specific architecture and/or
  machine type through the new 'require.arch' and 'require.machine'
  meta-data properties, respectively.

* Added the 'timeout' property to test cases, useful to set an upper-bound
  limit for the test's run time and thus prevent global test program stalls
  due to the test case's misbehavior.

* Fixed a serious bug that prevented atf-run(1) from working at all under
  Fedora 8 x86_64.  Due to the nature of the bug, other platforms were
  likely affected too.

* Added a new internal utility, atf-kill-tree(1), used to kill a tree of
  processes.


Version 0.3
===========

Release date: November 11th, 2007
Status:       Experimental

Changes:

* Added XML output support to atf-report.  This is accompanied by a DTD for
  the format's structure and sample XSLT/CSS files to post-process this
  output and convert it to a plain HTML report.

* Changed atf-run to add system information to the report it generates.
  This is currently used by atf-report's XML output only, and is later
  printed in the HTML reports in a nice and useful summary table.  The user
  and system administrator are allowed to tune this feature by means of
  hooks.

* Removed the test cases' 'isolated' property.  This was intended to avoid
  touching the file system at all when running the related test case, but
  this has not been true for a long while: some control files are
  unconditionally required for several purposes, and we cannot easily get
  rid of them.  This way we remove several critical and delicate pieces of
  code.

* Improved atf-report's CSV output format to include information about
  test programs too.

* Fixed the tests that used atf-compile to not require this tool as a
  helper.  Avoids systems without build-time utilities to skip many tests
  that could otherwise be run.  (E.g. NetBSD without the comp.tgz set
  installed.)

* Many general cleanups: Fixed many pieces of code marked as ugly and/or
  incomplete.


Version 0.2
===========

Release date: September 20th, 2007
Status:       Experimental

Changes:

* Test cases now get a known umask on entry.

* atf-run now detects many unexpected failures caused by test programs
  and reports them as bogus tests.  atf-report is able to handle these
  new errors and nicely reports them to the user.

* All the data formats read and written by the tools have been documented
  and cleaned up.  These include those grammars that define how the
  different components communicate with each other as well as the format
  of files written by the developers and users: the Atffiles and the
  configuration files.

* Added the atf-version tool, a utility that displays information about the
  currently installed version of ATF.

* Test cases can now define an optional cleanup routine to undo their
  actions regardless of their exit status.

* atf-report now summarizes the list of failed (bogus) test programs when
  using the ticker output format.

* Test programs now capture some termination signals and clean up any
  temporary files before exiting the program.

* Multiple bug fixes and improvements all around.


Version 0.1
===========

Release date: August 20th, 2007
Status:       Experimental

Changes:

* First public version.  This was released coinciding with the end of the
  Google Summer of Code 2007 program.


---------------------------------------------------------------------------
                              End of document


tools/atf-kill-tree.cpp
//
// Automated Testing Framework (atf)
//
// Copyright (c) 2008 The NetBSD Foundation, Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
// 3. All advertising materials mentioning features or use of this
//    software must display the following acknowledgement:
//        This product includes software developed by the NetBSD
//        Foundation, Inc. and its contributors.
// 4. Neither the name of The NetBSD Foundation nor the names of its
//    contributors may be used to endorse or promote products derived
//    from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND
// CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
// GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
// IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

extern "C" {
#include <signal.h>
}

#include <cstdlib>
#include <iostream>

#include "atf/application.hpp"
#include "atf/procs.hpp"
#include "atf/sanity.hpp"
#include "atf/text.hpp"
#include "atf/ui.hpp"

class atf_kill_tree : public atf::application::app {
    static const char* m_description;

    int m_signo;

    std::string specific_args(void) const;
    options_set specific_options(void) const;
    void process_option(int, const char*);

public:
    atf_kill_tree(void);

    int main(void);
};

const char* atf_kill_tree::m_description =
    "atf-kill-tree recursively kills a process and all of its "
    "children.";

atf_kill_tree::atf_kill_tree(void) :
    app(m_description, "atf-kill-tree(1)", "atf(7)"),
    m_signo(SIGTERM)
{
}

std::string
atf_kill_tree::specific_args(void)
    const
{
    return "pid";
}

atf_kill_tree::options_set
atf_kill_tree::specific_options(void)
    const
{
    using atf::application::option;
    options_set opts;
    opts.insert(option('s', "signal", "The signal to send (default SIGTERM)"));
    return opts;
}

void
atf_kill_tree::process_option(int ch, const char* arg)
{
    switch (ch) {
    case 's':
        m_signo = atf::text::to_type< int >(arg);
        break;

    default:
        UNREACHABLE;
    }
}

int
atf_kill_tree::main(void)
{
    if (m_argc == 0)
        throw atf::application::usage_error("Missing process identifier");

    pid_t pid = atf::text::to_type< pid_t >(m_argv[0]);

    atf::procs::pid_grabber pg;
    atf::procs::errors_vector errors = kill_tree(pid, m_signo, pg);
    for (atf::procs::errors_vector::const_iterator iter = errors.begin();
         iter != errors.end(); iter++) {
        pid_t epid = (*iter).first;
        const std::string& emsg = (*iter).second;

        std::string msg = "PID " + atf::text::to_string(epid) + ": " + emsg;
        std::cerr << atf::ui::format_error(m_prog_name, msg) << std::endl;
    }
    return errors.empty() ? EXIT_SUCCESS : EXIT_FAILURE;
}

int
main(int argc, char* const* argv)
{
    return atf_kill_tree().run(argc, argv);
}
